src/StackExchange.Redis/RedisSubscriber.cs

```diff
@@ -35,273 +55,178 @@ internal bool GetSubscriberCounts(in RedisChannel channel, out int handlers, out
-                }
+                completable = sub.ForInvoke(channel, payload, out queues);
+            }
+            if (queues != null)
+            {
+                ChannelMessageQueue.WriteAll(ref queues, channel, payload);
+            }
+            if (completable != null && !completable.TryComplete(false))
+            {
+                CompleteAsWorker(completable);
             }
-            if (queues != null) ChannelMessageQueue.WriteAll(ref queues, channel, payload);
-            if (completable != null && !completable.TryComplete(false)) CompleteAsWorker(completable);
         }
 
-        internal Task RemoveAllSubscriptions(CommandFlags flags, object asyncState)
+        /// <summary>
+        /// Updates all subscriptions re-evaluating their state.
+        /// This clears the current server if it's not connected, prepping them to reconnect.
+        /// </summary>
+        internal void UpdateSubscriptions()
         {
-            Task last = null;
-            lock (subscriptions)
+            foreach (var pair in subscriptions)
             {
-                foreach (var pair in subscriptions)
-                {
-                    pair.Value.MarkCompleted();
-                    var task = pair.Value.UnsubscribeFromServer(pair.Key, flags, asyncState, false);
-                    if (task != null) last = task;
-                }
-                subscriptions.Clear();
+                pair.Value.UpdateServer();
             }
-            return last ?? CompletedTask<bool>.Default(asyncState);
         }
 
-        internal Task RemoveSubscription(in RedisChannel channel, Action<RedisChannel, RedisValue> handler, ChannelMessageQueue queue, CommandFlags flags, object asyncState)
+        /// <summary>
+        /// Ensures all subscriptions are connected to a server, if possible.
+        /// </summary>
+        internal void EnsureSubscriptions(CommandFlags flags = CommandFlags.None)
         {
-            Task task = null;
-            lock (subscriptions)
+            // TODO: Subscribe with variadic commands to reduce round trips
+            foreach (var pair in subscriptions)
             {
-                if (subscriptions.TryGetValue(channel, out Subscription sub))
-                {
-                    bool remove;
-                    if (handler == null & queue == null) // blanket wipe
-                    {
-                        sub.MarkCompleted();
-                        remove = true;
-                    }
-                    else
-                    {
-                        remove = sub.Remove(handler, queue);
-                    }
-                    if (remove)
-                    {
-                        subscriptions.Remove(channel);
-                        task = sub.UnsubscribeFromServer(channel, flags, asyncState, false);
-                    }
-                }
+                DefaultSubscriber.EnsureSubscribedToServer(pair.Value, pair.Key, flags, true);
```

Comment:
This fires `n` times for `n` subscriptions, but is called with `FireAndForget` so is not O(n) blocking at least.